JARM for IS 5.10.0 (#3700)
wso2/docs-is: en/docs/learn/jarm-for-oauth2.md

# JWT Secured Authorization Response Mode (JARM) for OAuth 2.0

> The [JWT Secured Authorization Response Mode for OAuth 2.0 (JARM) specification](https://bitbucket.org/openid/fapi/src/master/oauth-v2-jarm.md) defines new JWT-based modes to encode OAuth2 authorization responses. These new modes allow clients to request authorization response parameters and additional data in JWT format.

Reviewer: https://openid.net/specs/openid-financial-api-jarm-ID1.html

Reviewer: lets use this link for the spec

> The [JWT Secured Authorization Response Mode for OAuth 2.0 (JARM) specification](https://bitbucket.org/openid/fapi/src/master/oauth-v2-jarm.md) defines new JWT-based modes to encode OAuth2 authorization responses. These new modes allow clients to request authorization response parameters and additional data in JWT format.

Reviewer: ```suggestion
The [JWT Secured Authorization Response Mode for OAuth 2.0 (JARM) specification](https://openid.net/specs/openid-financial-api-jarm-ID1.html) defines new JWT-based modes to encode OAuth2 authorization responses. These new modes allow clients to request authorization response parameters and additional data in JWT format.
```

!!! info
    This is only available as a WSO2 Update from WSO2 Identity Server update level **5.10.216** onwards. See the instructions on [updating WSO2 products](https://updates.docs.wso2.com/en/latest/).

## Authorization flow

Given below is a sample authorization request sent to the authorization endpoint of WSO2 Identity Server.

```bash

> https://<IS_HOME>/oauth2/authorize?prompt=login&scope=openid&redirect_uri=https://<CLIENT-HOST>/redirects/redirect1&client_id=<CLIENT-ID>&response_type=<RESPONSE-TYPE>&response_mode=<RESPONSE-MODE>

Reviewer 2: ```suggestion
https://<IS_HOME>/oauth2/authorize?prompt=login&scope=openid&redirect_uri=https://<CLIENT_HOST>/redirects/redirect1&client_id=<CLIENT_ID>&response_type=<RESPONSE_TYPE>&response_mode=<RESPONSE_MODE>
```

Shall we replace "-" with "_" for consistency? There are a few other occurrences as well.

            Possible values:
            <ul>
                <li>
                    <code>code</code>: Used if the <a href="../authorization-code-grant">authorization code flow</a> is implemented.
                </li>
                <li>
                    <code>token</code>: Used if the <a href="../implicit-grant">implicit flow</a> is implemented.
                </li>
                <li>
                    <code>code id_token</code>: Used if the <a href="../openid-connect-hybrid-flow">hybrid flow</a> is implemented.
                </li>
            </ul>
        </td>
    </tr>
    <tr>
        <th>
            response_mode
        </th>
        <td>
            Specifies how the authorization information should be returned to the client. By default, the following response modes are supported in WSO2 Identity Server: <code>query</code>, <code>fragment</code>, and <code>form_post</code>.</br></br>
            In the default authorization flow (for default response modes), the authorization response parameters, including the authorization code and ID token, are sent as human-readable plain text to the redirect URL, as shown below.</br></br>

>             <code>https://{CLIENT_HOST}/redirects/redirect1?code={AUTH-CODE}&session_state={SESSION-STATE}</code></br></br>

Reviewer 2: ```suggestion
            <code>https://<CLIENT_HOST>/redirects/redirect1?code=<AUTH_CODE>&session_state=<SESSION_STATE></code></br></br>
```

Author: unfortunately, we can't render the `<` syntax inside an html table.
But let me try and find a way to make this more consistent.

Reviewer 2: We can use &lt; instead of < and &gt; instead of >.
This should work:
<code>https://&lt;CLIENT_HOST&gt;/redirects/redirect1?code=&lt;AUTH-CODE&gt;&session_state=&lt;SESSION-STATE&gt;</code>

Author: this is done.

>             To change this response to a more secure JWT format, you need to use JARM.

Reviewer: ```suggestion
            The JARM specification focuses on sending this response in a more secure manner by encoding the response parameters in a jwt.
```

```

### fragment.jwt

Response parameters are encoded in a single JWT and sent as a fragment parameter to the redirect URI.

Sample response:

```bash
HTTP/1.1 302 Found
Location: https://<CLIENT_HOST>/redirects/redirect1#response=<JWT>
```

### form_post.jwt

The response parameter containing the JWT is encoded as an HTML form value that is auto-submitted in the User Agent, and thus is transmitted via the HTTP POST method to the client, with the result parameters being encoded in the body.

Sample response from the authorization server to the user agent:

```bash
HTTP/1.1 200 OK
Content-Type: text/html;charset=UTF-8
Cache-Control: no-cache, no-store
Pragma: no-cache

<html>
<head><title>Submit Form</title>
</head>
<body onload="javascript:document.forms[0].submit()">
<form method="post" action="https://<CLIENT_HOST>/redirects/redirect1">
    <input type="hidden" name="response" value=<JWT>/>
</form>
</body>
</html>
```

The above response results in the following POST request to the client's redirect URI:

```bash
POST /redirects/redirect1 HTTP/1.1
Host: <CLIENT_HOST>
Content-Type: application/x-www-form-urlencoded

response=<JWT>
```

### jwt

> The response mode jwt is a shortcut and indicates the default redirect encoding (query, fragment) for the requested response type, as shown below.

Reviewer: ```suggestion
The response mode jwt is a shortcut and indicates the default redirect encoding (query.jwt or fragment.jwt) for the requested response type, as shown below.
```

> - If <code>response_type</code> is <code>code</code>, the <code>response_mode</code> is <code>query.jwt</code>.

Reviewer: ```suggestion
- If <code>response_type</code> is <code>code</code>, then the default <code>response_mode</code> is <code>query.jwt</code>.
```

> - If <code>response_type</code> is <code>token</code>, the <code>response_mode</code> is <code>fragment.jwt</code>.

Reviewer: ```suggestion
- If <code>response_type</code> is <code>token</code>, then the default <code>response_mode</code> is <code>fragment.jwt</code>.
```

> - If <code>response_type</code> is <code>token</code>, the <code>response_mode</code> is <code>fragment.jwt</code>.

Reviewer: ```suggestion
```

> - If <code>response_type</code> is <code>token</code>, the <code>response_mode</code> is <code>fragment.jwt</code>.

Reviewer: ```suggestion

```

Reviewer: remove

> - If <code>response_type</code> contains <code>token</code> or <code>id_token</code>, the <code>response_mode</code> is <code>fragment.jwt</code>.

Reviewer: ```suggestion
- If <code>response_type</code> contains <code>token</code> or <code>id_token</code>, then the dafault <code>response_mode</code> is <code>fragment.jwt</code>.
```

> - If <code>response_type</code> contains <code>token</code> or <code>id_token</code>, the <code>response_mode</code> is <code>fragment.jwt</code>.

Reviewer: ```suggestion

```

Reviewer: remove

> 

Reviewer: - For the response types defined in [OIDM](https://openid.net/specs/oauth-v2-multiple-response-types-1_0.html) (​​except "none") the default response mode is <code>fragment.jwt</code>
(i.e. for response types containing <code>token</code> or <code>id_token</code> the default response mode is <code>fragment.jwt</code>)

Sample response from the authorization server to the user agent:

```bash
HTTP/1.1 200 OK
Content-Type: text/html;charset=UTF-8
Cache-Control: no-cache, no-store
Pragma: no-cache

<html>
<head><title>Submit Form</title>
</head>
<body onload="javascript:document.forms[0].submit()">
<form method="post" action="https://<CLIENT_HOST>/redirects/redirect1">
    <input type="hidden" name="response" value=<JWT>/>
</form>
</body>
</html>
```

The above response results in the following POST request to the client's redirect URI:

```bash
POST /redirects/redirect1 HTTP/1.1
Host: <CLIENT_HOST>
Content-Type: application/x-www-form-urlencoded

response=<JWT>
```

### jwt

The `jwt` response mode is a shortcut and it indicates the default redirect encoding (`query.jwt` or `fragment.jwt`) for the requested response type. That is, if the response mode is `jwt`, the server decides whether `query.jwt` or `fragment.jwt` is to be used depending on the `response_type` sent in the request.

The default response modes for each response type is shown below.

> - If the `response_type` is `code`, the `response_mode` is `query.jwt`.

Reviewer: ```suggestion
- If the `response_type` is `code`, then the default `response_mode` is `query.jwt`.
```
